tools/find-inactive-collaborators.mjs

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env node
+
+// Identify inactive collaborators. "Inactive" is not quite right, as the things
+// this checks for are not the entirety of collaborator activities. Still, it is
+// a pretty good proxy. Feel free to suggest or implement further metrics.
+
+import cp from 'node:child_process';
+import fs from 'node:fs';
+import readline from 'node:readline';
+
+const SINCE = process.argv[2] || '6 months ago';
+
+async function runGitCommand(cmd, mapFn) {
+  const childProcess = cp.spawn('/bin/sh', ['-c', cmd], {
+    cwd: new URL('..', import.meta.url),
+    encoding: 'utf8',
+    stdio: ['inherit', 'pipe', 'inherit'],
+  });
+  const lines = readline.createInterface({
+    input: childProcess.stdout,
+  });
+  const errorHandler = new Promise(
+    (_, reject) => childProcess.on('error', reject)
+  );
+  const returnedSet = new Set();
+  await Promise.race([errorHandler, Promise.resolve()]);
+  for await (const line of lines) {
+    await Promise.race([errorHandler, Promise.resolve()]);
+    const val = mapFn(line);
+    if (val) {
+      returnedSet.add(val);
+    }
+  }
+  return Promise.race([errorHandler, Promise.resolve(returnedSet)]);
+}
+
+// Retrieve all commit authors during the time period.
+const authors = await runGitCommand(
+  `git shortlog -n -s --since="${SINCE}"`,
+  (line) => line.trim().split('\t', 2)[1]
+);
+
+// Retrieve all commit landers during the time period.
+const landers = await runGitCommand(
+  `git shortlog -n -s -c --since="${SINCE}"`,
+  (line) => line.trim().split('\t', 2)[1]
+);
+
+// Retrieve all approving reviewers of landed commits during the time period.
+const approvingReviewers = await runGitCommand(
+  `git log --since="${SINCE}" | egrep "^    Reviewed-By: "`,
+  (line) => /^    Reviewed-By: ([^<]+)/.exec(line)[1].trim()
+);
+
+// Retrieve list of current collaborators from README.md.
+const readmeText = fs.readFileSync(
+  new URL('../README.md', import.meta.url),
+  'utf8'
+);
+let processingCollaborators = false;
+const collaborators = readmeText
+  .split('\n')
+  .filter((line) => {
+    const isCollaborator = processingCollaborators && line.length;
+    if (line === '### Collaborators') {
+      processingCollaborators = true;
+    }
+    if (line === '### Collaborator emeriti') {
+      processingCollaborators = false;
+    }
+    return line.startsWith('**') && isCollaborator;
+  })
+  .map((line) => line.split('**')[1].trim());
```

Comment:
Non-blocking suggestion: run everything in parallel? (I also added the `--no-pager` git flag, not sure if it's actually useful, but certainly can't hurt 🤷‍♂️)
```suggestion
async function retrieveCollaboratorsFromReadme() {
  const readmeText = readline.createInterface({
    input: fs.createReadStream(new URL('../README.md', import.meta.url)),
    crlfDelay: Infinity,
  });
  const returnedArray = [];
  let processingCollaborators = false;
  for await (const line of readmeText) {
    const isCollaborator = processingCollaborators && line.length;
    if (line === '### Collaborators') {
      processingCollaborators = true;
    }
    if (line === '### Collaborator emeriti') {
      processingCollaborators = false;
      break;
    }
    if (line.startsWith('**') && isCollaborator) {
      returnedArray.push(line.split('**', 2)[1].trim());
    }
  }
  return returnedArray;
}

const [authors, landers, approvingReviewers, collaborators] = await Promise.all(
  [
    // Retrieve all commit authors during the time period.
    exec(
      `git --no-pager shortlog -n -s --since="${SINCE}"`,
      (line) => line.trim().split('\t', 2)[1]
    ),

    // Retrieve all commit landers during the time period.
    exec(
      `git --no-pager shortlog -n -s -c --since="${SINCE}"`,
      (line) => line.trim().split('\t', 2)[1]
    ),

    // Retrieve all approving reviewers of landed commits during the time period.
    exec(
      `git --no-pager log --since="${SINCE}" | egrep "^    Reviewed-By: "`,
      (line) => /^    Reviewed-By: ([^<]+)/.exec(line)[1].trim()
    ),

    // Retrieve list of current collaborators from README.md.
    retrieveCollaboratorsFromReadme(),
  ]
);
```

Reply:
The Promise.all() overhead still makes this slower in all cases I've tested. Passing `'30 years ago'` as the argument takes 2.5 seconds to run in the current version but over 3 seconds with this change.

Reply:
I did go ahead and make the README collaborator thing async. That didn't affect performance measurably one way or the other.